errai-bus/src/main/java/org/jboss/errai/bus/server/service/ErraiServiceSingleton.java

```diff
@@ -30,29 +30,33 @@ private ErraiServiceSingleton() {
   private static Set<ErraiInitCallback> callbackList = Collections.synchronizedSet(new HashSet<ErraiInitCallback>());
 
   private static final Object monitor = new Object();
-  private static volatile ErraiService service = new ErraiServiceProxy();
+  private static volatile ErraiServiceProxy initProxy = new ErraiServiceProxy();
+  private static volatile ErraiService service;
 
   public static ErraiService initSingleton(final ErraiServiceConfigurator configurator) {
     synchronized (monitor) {
       if (isInitialized()) throw new IllegalStateException("service already set into singleton");
-      ErraiServiceProxy proxy = (ErraiServiceProxy) service;
+
       service = ErraiServiceFactory.create(configurator);
-      proxy.closeProxy(service);
+      if(!initProxy.isClosed()) {
+        initProxy.closeProxy(service);
+      }
 
       for (ErraiInitCallback erraiInitCallback : callbackList) {
         erraiInitCallback.onInit(service);
       }
 
+      callbackList.clear();
```

Comment:
We could do it that way I guess. It doesn't work well for me since a re-initialization will only occur when the servlets are refreshing so I need to add fresh callbacks in-case they have changed upon recomilation and refreshing. I will need to grab them and flush them myself if that is the case.

Reply:
Well, I just checked and this callback API is not used internally so there is nothing to break there either way. And since restarting the service was never part of the API before, I'd say do whatever works for your use case since it is the only use-case we have so far for restarting the service.

The only thing I ask here is that you add some javadoc to the register method explaining what the contract is.